### PR TITLE
[CI:BUILD] followup PR for fcos with podman-next

### PR DESCRIPTION
**Pull request by lsm5**
Followup on #19477

[NO NEW TESTS NEEDED]

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by mheon** [comment]
The Cirrus linter is printing warnings

**Comment by lsm5** [comment]
> The Cirrus linter is printing warnings

@mheon sorry, ignore this PR for now. It's just a placeholder. Thanks

**Comment by lsm5** [comment]
@benoitf btw, I hope `quay.io/podman/fcos:podman-next` has not been announced anywhere yet. I'm running a lot of the image build tests on https://github.com/containers/automation_sandbox so the SHA image tags are almost certainly messed up for now. Once this PR is merged, I'll be deleting prior SHA tags from quay.

**Comment by benoitf** [comment]
@lsm5 it's not advertised on podman desktop documentation. Some podman desktop developers will probably try ( for the one using macOS but it's ok if it's totally broken :-)

**Comment by lsm5** [comment]
@cevich I replaced the cirrus fcos build task with another github action. GHA does provide some reusable workflows but for now the pre-merge action is simple enough to not worry about reusability. In the pre-merge action I only care about using the Containerfile as-is along with pre-existing rpms from podman-next and not worry about installing the pre-merge rpms built by packit on the PR.

**Comment by lsm5** [comment]
this is what an image build and push action will look like after merge: https://github.com/containers/automation_sandbox/actions/runs/6113932428/job/16594431532 . Let's ensure it looks ok before merging.

**Comment by lsm5** [comment]
@cevich PTAL . The `Build FCOS image` still shows some errors about adding tag to manifest, but doesn't seem to be a deal-breaker https://github.com/containers/automation_sandbox/actions/runs/6123537430/job/16621629075 . quay images are pushed just fine .

File: .github/workflows/fcos-podman-next-build.yml
Comment (cevich) [inline .github/workflows/fcos-podman-next-build.yml:25]: IMHO, these two commands (esp. the first one) will introduce a non-trivial amount of  unpredictability in this workflow compared to `fcos-podman-next-build-prepush-test.yml`.  Would you consider doing this step, and the multi-arch build there as well?

The thing I'm afraid of, is the ability to notice a broken workflow early - within a PR,  rather than after merge.  From that perspective it would be good to keep the two workflows as identical as possible.

I'd even suggest (maybe for a future update), usinga local "reusable workflow" to de-duplicate at least the build portion.  There's [a simple example of doing this in the `podman.io` repo.](https://github.com/containers/podman.io/tree/636ba42ba742d05fdfb7090631d42e0f8d4bdf79/.github/workflows) - both `deploy.yml` and `test.yml` reuse `setup_and_build.yml`.

File: .github/workflows/fcos-podman-next-build.yml
Comment (cevich) [inline .github/workflows/fcos-podman-next-build.yml:50]: Minor: Tagging the image with a full sha still tickles my funny-bone slightly.  Mostly because in quay (or podman pull), it's not entirely obvious what the sha pertains to.  For example, it could be a digest from another image, a hash of some file, etc.  Would you consider using a prefixed tag instead, maybe something like `podman-${{ env.SHORTSHA }}`?  

Even better would be to (somehow) grab the V-R from the RPM's NVR and use that as the tag (along with `env.SHORTSHA`).

File: .github/workflows/fcos-podman-next-build.yml
Comment (cevich) [inline .github/workflows/fcos-podman-next-build.yml:58]: Thanks, this is helpful.  GH doesn't make it easy to observe these things by default.  Thanks for adding this.

File: .github/workflows/fcos-podman-next-build.yml
Comment (cevich) [inline .github/workflows/fcos-podman-next-build.yml:70]: Very minor: Skopeo would probably be faster for this.  I'd also suggest dumping the output into a file, then inspecting the file with `jq` rather than running the `inspect` command over and over.  All very minor issues though.  Feel free to ignore.

File: .github/workflows/fcos-podman-next-build.yml
Comment (cevich) [inline .github/workflows/fcos-podman-next-build.yml:73]: Glad to see this verification.  Would you consider actually running podman in the container instead of just bash?

```suggestion
      run: podman run --privileged --rm ${{ steps.build_image_multiarch.outputs.image-with-tag }} podman system info
```

**Review by cevich — COMMENTED** [review]
Really nice work Lokesh.  It mostly LGTM and likely plenty good enough.  I'm okay if you'd prefer to do the more important suggestions in a followup PR, if you'd prefer.  Also, thanks for taking the time to verify the workflows in the test-repo. as well.  I'm sure you uncovered some surprises and were able to fix them :wink:

File: .github/workflows/fcos-podman-next-build.yml
Comment (lsm5) [reply .github/workflows/fcos-podman-next-build.yml:25]: multi-arch build takes an awful long and it might end up being an annoyance if we do it before merge. I haven't yet checked if it takes longer than the rest of podman tests combined.

File: .github/workflows/fcos-podman-next-build.yml
Comment (lsm5) [reply .github/workflows/fcos-podman-next-build.yml:50]: Getting version-release in github action is probably convoluted. I'm gonna leave that for later. But `podman-shortsha` would be doable. @benoitf would you be ok with `podman-SHORTSHA` as a quay tag instead of the full sha ?

Comment (benoitf) [reply .github/workflows/fcos-podman-next-build.yml:50]: sure short sha works `podman-shortsha` or `git-shortsha` as well

File: .github/workflows/fcos-podman-next-build.yml
Comment (lsm5) [reply .github/workflows/fcos-podman-next-build.yml:73]: let me try this in sandbox.

File: .github/workflows/fcos-podman-next-build.yml
Comment (cevich) [reply .github/workflows/fcos-podman-next-build.yml:24]: Yes, the GHA VMs are slow *AND* the emulation is slow as well.  This was one of the reasons I did this in Cirrus, since we can simply "throw more hardware" at the problem.  I'm not sure that's a possible solution here w/o TONs of more work securing it all.

File: .github/workflows/fcos-podman-next-build.yml
Comment (lsm5) [reply .github/workflows/fcos-podman-next-build.yml]: this works. Added to PR.

File: .github/workflows/fcos-podman-next-build.yml
Comment (lsm5) [reply .github/workflows/fcos-podman-next-build.yml]: skopeo inspect failed, see: https://github.com/containers/automation_sandbox/actions/runs/6163597314/job/16727478272 . I guess we can leave skopeo switch to a followup PR.

File: .github/workflows/fcos-podman-next-build.yml
Comment (lsm5) [reply .github/workflows/fcos-podman-next-build.yml]: the github action now says `podman-SHORTSHA`. A sample image is already available at https://quay.io/repository/podman/fcos?tab=tags in that format.

File: .github/workflows/fcos-podman-next-build.yml
Comment (lsm5) [reply .github/workflows/fcos-podman-next-build.yml:25]: ack, let's leave any possible reusability for another followup later on. I can add comments in the actions file.

**Comment by lsm5** [comment]
Still need to check if I can add annotations.

**Comment by lsm5** [comment]
> Still need to check if I can add annotations.

will try this in sandbox first.

**Comment by vrothberg** [comment]
> @mheon sorry, ignore this PR for now. It's just a placeholder. Thanks

Maybe convert the PR into a draft, so it's more obvious that it's still WIP and it will prevent us from accidentally merging.

**Comment by lsm5** [comment]
@benoitf would you like `crun-wasm` pulled into these fcos images? I guess I could do that in a second followup PR. I'd rather not rerun all tests here since this PR should be good to go.

**Comment by benoitf** [comment]
ah yes that would be awesome

`crun-wasm` and `wasmedge-rt`

and yes another PR will be fine

**Review by cevich — APPROVED** [review]
/lgtm
/hold

**Comment by lsm5** [comment]
the github action failed at `wait-for-copr` https://github.com/containers/podman/actions/runs/6247273398/job/16959529269#step:6:16 . Seems like a packit / copr issue. The packit team is investigating this. Once they have it fixed, our GHA should work as well.

**Comment by lsm5** [comment]
ugh, the changes in `.packit.sh` for `podman --version` to show short_sha also don't seem to have taken effect, don't know why. Will work on that in yet another followup.

**Comment by cevich** [comment]
> the github action failed at wait-for-copr

When it's fixed, it might be worth considering if you should pin onto that working version somehow.  The caution here is, don't have any hope of managing updates to that with Renovate, since wait-for-copr isn't tested in the PR workflow.

> ugh, the changes in .packit.sh for podman --version to show short_sha also don't seem to have taken effect

Oh man, sorry :cry: Let me guess, that isn't/wasn't something you could test in the sandbox?

> Will work on that in yet another followup.

These kinds of problems are only amplified by not having full end-to-end testing in PRs.  Not sure how that could be done in this case though.  At least easily.

-sigh-

**Comment by lsm5** [comment]
> When it's fixed, it might be worth considering if you should pin onto that working version somehow. The caution here is, don't have any hope of managing updates to that with Renovate, since wait-for-copr isn't tested in the PR workflow.

ya, pinning will work. I can change that once it's fixed. And I guess renovate only works on go.mod. No use if the repo is only mentioned in GHA.

> Oh man, sorry 😢 Let me guess, that isn't/wasn't something you could test in the sandbox?

> These kinds of problems are only amplified by not having full end-to-end testing in PRs. Not sure how that could be done in this case though. At least easily.

What happened in this case was that I verified the short sha in `./bin/podman --version` by running make locally. I forgot to verify it using the rpms built by packit. So, this could have been noticed earlier had I remembered to look in the right place.

**Comment by lsm5** [comment]
> > When it's fixed, it might be worth considering if you should pin onto that working version somehow. The caution here is, don't have any hope of managing updates to that with Renovate, since wait-for-copr isn't tested in the PR workflow.
> 
> ya, pinning will work. I can change that once it's fixed. And I guess renovate only works on go.mod. No use if the repo is only mentioned in GHA.
> 

@cevich hmm, in this case if it was a problem in `copr` itself and not `wait-for-copr` there's only so much pinning would do I guess.

**Comment by cevich** [comment]
> And I guess renovate only works on go.mod. No use if the repo is only mentioned in GHA.

Actually it has a really powerful mechanism to manage versions using regular-expressions.  The problem here is any update PR it proposed wouldn't actually run the thing being updated. 
So nobody would know if the update itself broke something.

> there's only so much pinning would do I guess.

It would prevent an upstream breaking-change (or broken change) from affecting automation.

**Comment by lsm5** [comment]
> > And I guess renovate only works on go.mod. No use if the repo is only mentioned in GHA.
> 
> Actually it has a really powerful mechanism to manage versions using regular-expressions. The problem here is any update PR it proposed wouldn't actually run the thing being updated. So nobody would know if the update itself broke something.
> 
> > there's only so much pinning would do I guess.
> 
> It would prevent an upstream breaking-change (or broken change) from affecting automation.

I think I can do this once the issue is resolved on the copr side. So far, we've only had 2 successful image builds using wait-for-copr.

**Pull request merged**
